src/model/enhance-images.ts

```diff
@@ -0,0 +1,333 @@
+ * @param {Boolean} isPhotoEssay - We treat photo essays differently
+ * */
+const decideTitle = ({
+	element,
+	isPhotoEssay,
+}: {
+	element?: SubheadingBlockElement;
+	isPhotoEssay: boolean;
+}): string => {
+	// Checks if this element is a 'title' based on the convention: <h2>Title text</h2>
+	if (!element) return '';
+	if (!isPhotoEssay) return '';
+	// Extract title
+	const frag = JSDOM.fragment(element.html);
+	if (!frag || !frag.firstChild) return '';
+	const isH2tag = frag.firstChild.nodeName === 'H2';
+	if (isH2tag) {
+		// element is an essay title
+		return (frag.textContent && frag.textContent.trim()) || '';
+	}
+	return '';
+};
+
+const enhanceImage = ({
+	image,
+	caption,
+	title,
+}: {
+	image: ImageBlockElement;
+	caption?: string;
+	title?: string;
+}): ImageBlockElement => {
+	const enhancedImage = { ...image };
+	if (caption) {
+		enhancedImage.data.caption = caption;
+	}
+	if (title) {
+		enhancedImage.title = title;
+	}
+	return enhancedImage;
+};
+
+const processBuffer = ({
+	imageBuffer,
+	title,
+	caption,
+	isPhotoEssay,
+}: {
+	imageBuffer: ImageBlockElement[];
+	title?: string;
+	caption?: string;
+	isPhotoEssay?: boolean;
+}): CAPIElement[] => {
+	function cleanCaption(image: ImageBlockElement) {
+		// If a standalone caption was found we want to strip any existing captions
+		// Or if this is article is a photo essay, then we always strip captions, even if that
+		// leaves the image with no caption at all, we want that for photo essays
+		if (caption || isPhotoEssay) {
+			return {
+				...image,
+				data: {
+					...image.data,
+					caption: '',
+				},
+				displayCredit: false,
+			};
+		}
+		// No special caption was found so continue with this images defaults
+		return image;
+	}
+
+	const processed: CAPIElement[] = [];
+	let prevHalfWidth: ImageBlockElement | null;
+	imageBuffer.map(cleanCaption).forEach((image, i) => {
+		const endOfBuffer = i + 1 === imageBuffer.length;
+		switch (image.role) {
+			case 'halfWidth':
+				if (!prevHalfWidth) {
+					if (endOfBuffer) {
+						processed.push(enhanceImage({ image, caption, title }));
+					} else {
+						prevHalfWidth = image;
+					}
+				} else {
+					if (endOfBuffer) {
+						processed.push({
+							_type:
+								'model.dotcomrendering.pageElements.MultiImageBlockElement',
+							elementId: prevHalfWidth.elementId,
+							images: [prevHalfWidth, image],
+							caption,
+						});
+					} else {
+						processed.push({
+							_type:
+								'model.dotcomrendering.pageElements.MultiImageBlockElement',
+							elementId: prevHalfWidth.elementId,
+							images: [prevHalfWidth, image],
+						});
+					}
+					prevHalfWidth = null;
+				}
+				break;
+			case 'inline':
+			case 'immersive':
+			case 'showcase':
+			case 'supporting':
+			case 'thumbnail':
+			default:
+				if (endOfBuffer) {
+					processed.push(enhanceImage({ image, caption, title }));
+					// Mop up any dangling halfWidth images that never had a sibling
+					if (prevHalfWidth)
```

Comment:
Might be worth thinking on the name of `prevHalfWidth` because I stumbled a couple of times, something like `prevHalfWidthImageBlockElement` (for brevity obvs...).